numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -401,143 +654,129 @@ init_comparison(PyObject *umath)
 }
 
 
-static NPY_CASTING
-string_addition_resolve_descriptors(
-        PyArrayMethodObject *NPY_UNUSED(self),
-        PyArray_DTypeMeta *NPY_UNUSED(dtypes[3]),
-        PyArray_Descr *given_descrs[3],
-        PyArray_Descr *loop_descrs[3],
-        npy_intp *NPY_UNUSED(view_offset))
-{
-    loop_descrs[0] = NPY_DT_CALL_ensure_canonical(given_descrs[0]);
-    if (loop_descrs[0] == NULL) {
-        return _NPY_ERROR_OCCURRED_IN_CAST;
-    }
-
-    loop_descrs[1] = NPY_DT_CALL_ensure_canonical(given_descrs[1]);
-    if (loop_descrs[1] == NULL) {
-        return _NPY_ERROR_OCCURRED_IN_CAST;
-    }
-
-    loop_descrs[2] = PyArray_DescrNew(loop_descrs[0]);
-    if (loop_descrs[2] == NULL) {
-        return _NPY_ERROR_OCCURRED_IN_CAST;
-    }
-    loop_descrs[2]->elsize += loop_descrs[1]->elsize;
-
-    return NPY_NO_CASTING;
-}
-
-
 static int
-init_add(PyObject *umath)
+init_ufunc(PyObject *umath, const char *name, const char *specname, int nin, int nout,
```

Comment:
I think it might be cleaner to add an `enc` template parameter that does different logic for the different encodings and only accept a single `PyArrayMethod_StridedLoop`. The caller would then call `init_ufunc` twice (eventually three times) for the different encodings.

Reply:
Agreed. Would it be okay for me to do this in another PR, after I've reworked `add` and `isalpha` to work with the new buffer class? Right now, they expect a different template parameter, so this wouldn't work very well and wouldn't be as clear.

Reply:
Sounds good!